src/shaders/pbr.frag

```diff
@@ -0,0 +1,246 @@
+  float n_dot_v = max(dot(normal, view), 0.0);
+  float ggx1 = geometrySchlickGGX(n_dot_l, roughness);
+  float ggx2 = geometrySchlickGGX(n_dot_v, roughness);
+
+  return ggx1 * ggx2;
+}
+
+// Specular F, aka Fresnel, use Schlick's approximation
+// F0: specular reflectance at normal incidence
+// view: camera direction, aka light outgoing direction
+// half: half vector of light and view
+vec3 fresnelSchlick(vec3 F0, vec3 view, vec3 half) {
+  float v_dot_h = max(dot(view, half), 0.0);
+  return F0 + (1.0 - F0) * pow(1.0 - v_dot_h, 5.0);
+}
+
+// albedoColor: diffuse color from albedoTexture
+//              or from baseColor in the material;
+// metallic: metalness of the surface, from metallicTexture
+//           or from metallic in the material;
+// roughness: roughness of the surface, from roughnessTexture
+//            or from roughness in the material;
+// normal: normal direction
+// light: light source direction
+// view: camera direction, aka light outgoing direction
+// lightRadiance: the radiance of the light,
+//                which equals to intensity * attenuation
+vec3 microfacetModel(vec3 albedoColor,
+                     float metallic,
+                     float roughness,
+                     vec3 normal,
+                     vec3 light,
+                     vec3 view,
+                     vec3 lightRadiance) {
+  vec3 half = normalize(l + v);
+  // compute F0, specular reflectance at normal incidence
+  // for nonmetal, using constant 0.04
+  vec3 F0 = mix(vec3(0.04), albedoColor, metallic);
+  vec3 Fresnel = fresnelSchlick(F0, view, half);
+
+  // Diffuse BRDF
+  // NOTE: energy conservation requires
+  // diffuse + specular <= 1.0, where specular = Fresnel
+  // Also: result does not need to be scaled by 1/PI
+  // See details:
+  // https://seblagarde.wordpress.com/2012/01/08/pi-or-not-to-pi-in-game-lighting-equation/
+  vec3 diffuse = mix(vec3(1.0) - Fresnel, vec3(0.0), metallic) * albedoColor;
+
+  // Specular BDDF
+  float temp =
+      max(4.0 * max(dot(normal, light), 0.0) * max(dot(normal, view), 0.0),
+          Epsilon);
+  vec3 specular = Fresnel *
+                  specularGeometricAttenuation(normal, light, view, roughness) *
+                  normalDistribution(normal, half, roughness) / temp;
+
+  return (diffuse + specular) * lightRadiance * max(dot(normal, light), 0.0);
+}
+
+void main() {
+#if defined(ALBEDO_TEXTURE)
+  vec4 albedoColor = texture(albedoTexture, texCoord);
+#else
+  vec4 albedoColor = Material.baseColor;
+#endif
+#if defined(METALLIC_TEXTURE)
+  float metallic = texture(metallicTexture, texCoord).r;
+#else
+  float metallic = Material.metallic;
+#endif
+#if defined(ROUGHNESS_TEXTURE)
+  float roughness = texture(roughnessTexture, texCoord).r;
+#else
+  float roughness = Material.roughness;
+#endif
+#if defined(NORMAL_TEXTURE)
+  // normal is now in the camera space
+  vec3 n = getNormalFromNormalMap();
+#else
+  vec3 n = normal;
+#endif
```

Comment:
Here you might want to include normal map scale as well -- glTF has that, the Phong shader [recently received support for it](https://doc.magnum.graphics/magnum/classMagnum_1_1Shaders_1_1Phong.html#ac20d44948e41e8371daebe198a7b9b99) and I think @eundersander started using it for some models (instead of scaling the texture data). [Details how is it applied are here.](https://doc.magnum.graphics/magnum/namespaceMagnum_1_1Trade.html#ac20668da810bce38bddd4f4278f93662ab59103e987d5c0e6e17c6c184edf90b2)